### ISSUE TITLE
feature/multigrid and develop compilation issue

### ISSUE THREAD

**Issue by cpviolator** [lattice/quda#508]
I'm getting a persistent error message when compiling feature/multigrid or develop branch with multigrid enabled. I'm using magma 1.6.2 compiled with 'gcc with MKL 10.3, Intel threads' which passes all tests. I have set `MAGMA_INC` and `MAGMA_LIB` as environment variables in my `.bashrc` file.

When compiling Quda, `NVCC` will throw the error:

`nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified`

when attempting to compile the `blas_magma.cu` file. Google has not been helpful on this error, except to warn that if a source file has spaces, it should be enclosed in quotation marks. I don't think that applies here.

Has anybody seen this error before, or have an idea of where I may be going wrong? I have attached my `make.inc` file for reference. Environment variables are:

`echo $MAGMA_INC 
/work/03653/howarth/magma-1.6.2/include`

`echo $MAGMA_LIB
/work/03653/howarth/magma-1.6.2/lib`

`GCC` is 4.7.1
`CUDA` is 6.5

[make.inc.txt](https://github.com/lattice/quda/files/487899/make.inc.txt)

**Comment by maddyscientist** [comment]
Hmm, haven't seen this issue before.

Can you try compiling with cmake instead of configure?  https://github.com/lattice/quda/wiki/Building-QUDA-with-cmake
(Building with configure is deprecated.)

Also, you should stick to develop and ignore feature/multigrid, as everything in feature/multigrid has already been merged into develop.

Lastly, probably best to upgrade to a more recent version of CUDA, e.g., 7.5.

**Comment by alexstrel** [comment]
to this topic,  unfortunately different versions of MAGMA are not compatible . I believe you changed line 14 in bias_magma.cu with MAGMA_14 (to support pre 1.7.X versions) 
Also I just realized: MAGMA 2.X.X versions are not supported as well, because they changed interface functions (again!). This should be fixed.

**Comment by cpviolator** [comment]
OK, now using CMAKE 3.1.0 with CUDA 7.5 and GCC 4.9.3. Magma was complied with intel MKL 

I did get one initial error with CMAKE where in `CMakeLists.txt` I had to replace:

`message( ${MAGMA_INCLUDEDIR})`
with
`message("${MAGMA_INCLUDEDIR}")`

to avoid a CMake error

Now, everything goes well until CMake gets to the tests. The error thrown is:

`[ 85%] Built target quda_test
Linking CXX executable blas_test
/opt/apps/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/../../../../x86_64-unknown-linux-gnu/bin/ld: libquda_test.a(test_util.cpp.o): undefined reference to symbol 'cabsf@@GLIBC_2.2.5'
/lib64/libm.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [tests/blas_test] Error 1
make[1]: *** [tests/CMakeFiles/blas_test.dir/all] Error 2
make: *** [all] Error 2`

so obviously it can't find the `libm.so.6` library. Is there some flag I'm missing in CMake? I know this isn't really a QUDA issue but I'm REALLY close to successful compilation and I'm totally stuck on this.
[CMakeCache.txt](https://github.com/lattice/quda/files/488461/CMakeCache.txt)

**Comment by cpviolator** [comment]
Just to close this off.

I chose to use `magma-1.7.0` with `gcc 4.9.3` and `gnu threads`, and I linked against `lib magma.so`. Complilation was successful, but only after I explicitly replaced

`#include <magma.h>`
with
`#include </explicit/path/to/my/magma/include/magma.h>`
in the `lib/blas_magma.cu` file.

Currently trying to run `multigrid_benchmark_test` but getting the error:
`ERROR: Unsupported number of coarse dof 8
 (/work/03653/howarth/quda-develop-22092016-0.8.0/lib/dslash_coarse.cu:655 in ApplyCoarse())
       last kernel called was (name=N4quda9CopyGaugeIffLi128ENS_5gauge10FieldOrderIfLi8ELi1EL21QudaGaugeFieldOrder_s2EEENS2_IfLi8ELi1ELS3_5EEELb0EEE,volume=24x24x24x24,aux=)`

If I can't solve it I'll create a new issue.

**Comment by maddyscientist** [comment]
Thanks for the update.  Was your successful compilation with cmake or configure?

The multigrid_benchmark_test is a pure perf benchmark, and indeed as you've found by default it won't run (unnecessary parts of the code are not instantiated to try and minimize compilation time).  You can make the multigrid_benchmark_test work by editing this file to only test Nc=24 and 32, e.g., set the initial loop value to be 24, and increment in steps of 8.

Regardless, the real test is the multigrid_invert_test; there's a wiki page about how to use this, but I need to update it as it's a bit out of date.

**Comment by cpviolator** [comment]
Yes, I used CMake. I was confused by the `MAGMA` option (I thought it was just the magma directory it wanted) but CMake threw enough info to work it out.

Will do RE the benchmark test. I will be doing a lot of MG comparisons with other solver anyway. `multigrid_invert_test` is working splendidly. Awesome work!

**Comment by maddyscientist** [comment]
The MG in QUDA has been my biggest labour of love / hate ever 😉 

There's a bunch of parameters in MG, if you want we can sync up on the best ones to use.  And I should really update the wiki!

One thought on MAGMA, it uses pkg-config ($(prefix)/lib/pkgconfig/magma.pc), I'm guessing something is going wrong with that which is resulting you needing to use absolute path for the include.

I plan on removing the MAGMA dependency anyway for MG, it was purely for convenience I used this.

**Comment by cpviolator** [comment]
Getting some optimal parameters would be great, thanks! I saw that you did some comparisons with `CG` and `BiCGstab` here #451 and I will be doing comparisons with a deflated solver for TM and TM clover. The eigensolver is a real time sink for TM because we must calculate with the normal operator M^{\dag}M (unlike \gamma_5 hermitian operators where \gamma_5 M can be passed to the eigenslover to get the SVD of M) so this should be a real boost to the calculations.

What will replace magma?

**Comment by maddyscientist** [comment]
CUBLAS / CUDENSE has everything we need for MG (batched LU solvers and batched matrix inverse), so we can just use that (it would take me a day to fix this, but I've had too many other things).  Fancy giving it a shot? 😜

**Comment by cpviolator** [comment]
HA! Maybe! I have a few jobs to do here at Temple too but I do enjoy a challenge. I assume that it's mostly just replacing the functions in `blas_magma.cu` with the appropriate ones in CUBLAS/CUDENSE? I'll take a look for sure.

**Comment by maddyscientist** [comment]
Yes, mostly.  MAGMA is called from dirac_coarse.cpp, you need to switch out that for the CUBLAS / CUDENSE variant.  Only minor compilation is that MAGMA provides helper functions for setting the pointers into the batched matrices, not sure if CUBLAS includes this too, or if you'll have to do that yourself.

Also, check the pm I just sent to your gmail.
